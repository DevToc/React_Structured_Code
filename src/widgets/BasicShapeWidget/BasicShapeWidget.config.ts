export const DEFAULT_WIDTH = 100;
export const DEFAULT_HEIGHT = 100;

export const DEFAULT_FILL_COLOR = '#0073E6';
export const DEFAULT_BORDER_WIDTH = 5;
export const DEFAULT_SECONDARY_FILL_COLOR = '#F64D2A';
