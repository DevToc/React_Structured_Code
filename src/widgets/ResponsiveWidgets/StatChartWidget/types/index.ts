import { StatChartType } from '../StatChartWidget.types';
import { DonutStatChart } from './DonutStatChart';
import { HalfDonutStatChart } from './HalfDonutStatChart';
import { ProgressBarStatChart } from './ProgressBarStatChart';
import { IconStatChart } from './IconStatChart';

export const CHART_TYPE_MAP = {
  [StatChartType.Donut]: DonutStatChart,
  [StatChartType.HalfDonut]: HalfDonutStatChart,
  [StatChartType.ProgressBar]: ProgressBarStatChart,
  [StatChartType.Icon]: IconStatChart,
};
