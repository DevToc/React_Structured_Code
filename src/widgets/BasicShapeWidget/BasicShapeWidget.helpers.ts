import { WidgetType } from '../../types/widget.types';
import { BasicShapeType, BasicShapeWidgetData, BorderStyle } from './BasicShapeWidget.types';
import { DEFAULT_WIDTH, DEFAULT_HEIGHT, DEFAULT_FILL_COLOR } from './BasicShapeWidget.config';
import { VERSION } from './BasicShapeWidget.upgrade';

const DEFAULT_BORDER = {
  border: {
    color: DEFAULT_FILL_COLOR,
    width: 5,
    style: BorderStyle.Solid,
  },
};

const TRANSPARENT_BORDER = {
  border: {
    color: 'black',
    width: 0,
    style: BorderStyle.Solid,
  },
};

export const generateShapeStyle = (mirror: { isVertical?: boolean; isHorizontal?: boolean }) => {
  const style = { position: 'absolute' as 'absolute' };

  const { isHorizontal, isVertical } = mirror || {};
  const scaleX = isHorizontal ? -1 : 1;
  const scaleY = isVertical ? -1 : 1;

  if (isVertical || isHorizontal) {
    return {
      ...style,
      transform: `scale(${scaleX}, ${scaleY})`,
      transformOrigin: 'center',
    };
  }
  return style;
};

/**
 * Generate default BasicShapeWidget data
 *
 * @param type {BasicShapeType} Shape type
 * @param isBorder if TRUE, will return border-only shape data (fill = 'none', border = default border), otherwise, will return fill-only data
 * @param options [optional] can specify any other attributes to be added to the widget data
 * @returns
 */
export const generateDefaultData = (
  type: BasicShapeType,
  isBorder = false,
  options: Partial<BasicShapeWidgetData> = {},
) => {
  const border = isBorder ? DEFAULT_BORDER : TRANSPARENT_BORDER;
  const fillColor = isBorder ? 'transparent' : DEFAULT_FILL_COLOR;

  return Object.assign({
    widgetType: WidgetType.BasicShape,
    widgetData: {
      version: VERSION,
      widthPx: DEFAULT_WIDTH,
      heightPx: DEFAULT_HEIGHT,
      topPx: 0,
      leftPx: 0,
      rotateDeg: 0,
      altText: '',
      isDecorative: true,

      // BasicShapeWidget properties
      fillColor: [fillColor],
      fillPercent: 0,
      mirror: {
        isHorizontal: false,
        isVertical: false,
      },
      type,
      ...border,
      ...(options && options),
    },
  });
};
