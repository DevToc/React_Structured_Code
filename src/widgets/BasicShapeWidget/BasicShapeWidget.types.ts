import { AccessibleElement, Widget } from 'types/widget.types';

export const enum BasicShapeType {
  Rectangle = 'rectangle',
  Ellipse = 'ellipse',
  SemiCircle = 'semi-circle',
  Triangle = 'triangle',
  RightTriangle = 'right-triangle',
}

export const enum BorderStyle {
  Dotted = 'dotted',
  Dashed = 'dashed',
  Solid = 'solid',
}

interface BasicShapeWidgetData extends Widget, AccessibleElement {
  type: BasicShapeType;
  border: {
    color: string;
    width: number;
    style: BorderStyle;
  };
  fillColor: string[];
  fillPercent: number;
  cornerRadius?: number;

  // generic icon data
  mirror: {
    isVertical?: boolean;
    isHorizontal?: boolean;
  };
}

interface BasicShapeLayoutProps {
  dimension: {
    width: number;
    height: number;
  };
  includeAltTextImg?: boolean;
}

interface BasicShapeToolbarMenuProps {
  // widgetId: WidgetId;
}

export type { BasicShapeWidgetData, BasicShapeLayoutProps, BasicShapeToolbarMenuProps };
