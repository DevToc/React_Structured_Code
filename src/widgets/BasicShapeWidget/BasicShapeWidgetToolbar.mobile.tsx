import { ReactElement, Suspense, useMemo } from 'react';

import {
  Button,
  Drawer,
  DrawerOverlay,
  DrawerContent,
  DrawerHeader,
  DrawerBody,
  Flex,
  IconButton,
  Box,
  Text,
  useDisclosure,
} from '@chakra-ui/react';

import { useWindowSize } from 'hooks/useWindowSize';

import { AltTextMenu } from 'modules/Editor/components/Toolbar/AltTextMenu/AltTextMenu';
import { BasicShapeWidgetData, BorderStyle } from './BasicShapeWidget.types';
import { DEFAULT_BORDER_WIDTH, DEFAULT_SECONDARY_FILL_COLOR } from './BasicShapeWidget.config';
import { ToolbarDivider } from 'modules/common/components/Toolbar/ToolbarDivider';
import { SliderNumberInput } from 'modules/common/components/Input/SliderNumberInput';
import { ColorPickerInContext } from 'modules/Editor/components/ColorPicker';
import { useWidget } from 'widgets/sdk';

import { ReactComponent as BorderStyleIcon } from 'assets/icons/border_style.svg';
import { ReactComponent as LineWidthIcon } from 'assets/icons/line_width.svg';
import { ReactComponent as HorizontalMirrorIcon } from 'assets/icons/mirror_horizontal.svg';
import { ReactComponent as VerticalMirrorIcon } from 'assets/icons/mirror_vertical.svg';
import { ReactComponent as SolidLineIcon } from 'assets/icons/solid_line.svg';
import { ReactComponent as DashedLineIcon } from 'assets/icons/dashed_line.svg';
import { ReactComponent as DottedLineIcon } from 'assets/icons/dotted_line.svg';
import { ReactComponent as FillIcon } from 'assets/icons/drop.svg';
import { ReactComponent as CloseIcon } from 'assets/icons/close_circle.svg';

const COLOR_ICON_SIZE = 5;
const TRIGGER_BUTTON_SIZE = 8;

const BORDER_STYLE_OPTIONS = [
  {
    value: BorderStyle.Solid,
    label: 'Solid',
    icon: <SolidLineIcon />,
  },
  {
    value: BorderStyle.Dashed,
    label: 'Dashed',
    icon: <DashedLineIcon />,
  },
  {
    value: BorderStyle.Dotted,
    label: 'Dotted',
    icon: <DottedLineIcon />,
  },
];

interface DefaultTriggerProps {
  onClick: () => void;
}
type IconStyle = 'fill' | 'border' | 'text';

interface Props extends DefaultTriggerProps {
  iconStyle: IconStyle;
  color: string;
}

export const BasicShapeWidgetToolbarMenuMobile = (): ReactElement => {
  const { deviceWidth } = useWindowSize();

  return (
    <>
      <Flex w={deviceWidth - 15} p='10px' gap='8px' align='center' data-testid='basic-shape-widget-toolbar'>
        <Fill />
        <BorderColorOption />

        <ToolbarDivider height={10} />
        <MirrorOptions />
        <ToolbarDivider height={10} />

        <BorderWidthOption />
        <BorderStyleOption />

        <ToolbarDivider height={10} />
        <AltTextMenuWrapper />
      </Flex>
    </>
  );
};

const AltTextMenuWrapper = () => {
  const { widgetId } = useWidget<BasicShapeWidgetData>();

  return (
    <Box //
      mr='5px'
      p='10px 10px'
      as='div'
      style={{ display: 'flex', flexDirection: 'column', alignItems: 'center' }}
    >
      <Box w='75px' as='div'>
        <AltTextMenu widgetId={widgetId} />
      </Box>
    </Box>
  );
};

const Fill = () => {
  const { fillColor, fillPercent, updateWidget } = useWidget<BasicShapeWidgetData>();
  const { isOpen: isColorFillOpen, onToggle: toggleColorFill } = useDisclosure();
  const { isOpen: isFillColorOpen, onToggle: toggleFillColor } = useDisclosure();
  const { isOpen: isFillColorTwoOpen, onToggle: toggleFillColorTwo } = useDisclosure();

  const colorFillTriggerProps: DefaultTriggerProps = useMemo(
    () => ({
      onClick: toggleFillColor,
    }),
    [toggleFillColor],
  );

  const colorFillTwoTriggerProps: DefaultTriggerProps = useMemo(
    () => ({
      onClick: toggleFillColorTwo,
    }),
    [toggleFillColorTwo],
  );

  const onUpdatePrimaryFillColor = (newFillColor: string): void => {
    const widgetFillColor = { fillColor: [newFillColor] };
    if (fillColor[1]) widgetFillColor.fillColor.push(fillColor[1]);

    updateWidget(widgetFillColor);
  };

  const onUpdateSecondaryFillColor = (newFillColor: string): void => {
    const widgetFillColor = { fillColor: [fillColor[0], newFillColor] };
    updateWidget(widgetFillColor);
  };

  const onUpdateFillPercent = (newFillPercent: number) => {
    const widgetData: { fillPercent: number; fillColor?: string[] } = { fillPercent: newFillPercent };
    if (fillPercent === 0 && fillColor.length === 1) {
      widgetData.fillColor = [...fillColor, DEFAULT_SECONDARY_FILL_COLOR];
    }

    updateWidget(widgetData);
  };

  const hasSecondaryColorFill = fillColor[1] && fillPercent && fillPercent > 0 && fillPercent <= 100;

  return (
    <>
      <Box
        onClick={() => toggleColorFill()}
        mr='5px'
        p='10px 10px'
        as='div'
        style={{ display: 'flex', flexDirection: 'column', alignItems: 'center' }}
      >
        <IconButton mb='10px' size='sm' aria-label='Font Icon' icon={<FillIcon />} />
        <Box w='75px' as='div'>
          <Text fontSize={10} style={{ fontWeight: 700 }} align='center'>
            Color Fill %
          </Text>
        </Box>
      </Box>

      <Box
        onClick={() => toggleFillColor()}
        mr='5px'
        p='10px 10px'
        as='div'
        style={{ display: 'flex', flexDirection: 'column', alignItems: 'center' }}
      >
        <DefaultTrigger iconStyle='fill' color={fillColor[0]} {...colorFillTriggerProps} />

        <Box w='75px' as='div'>
          <Text fontSize={10} style={{ fontWeight: 700 }} align='center'>
            Fill Color
          </Text>
        </Box>
      </Box>

      {hasSecondaryColorFill && (
        <Box
          onClick={() => toggleFillColorTwo()}
          mr='5px'
          p='10px 10px'
          as='div'
          style={{ display: 'flex', flexDirection: 'column', alignItems: 'center' }}
        >
          <DefaultTrigger iconStyle='fill' color={fillColor[1]} {...colorFillTwoTriggerProps} />

          <Box w='75px' as='div'>
            <Text fontSize={10} style={{ fontWeight: 700 }} align='center'>
              Fill Color Two
            </Text>
          </Box>
        </Box>
      )}

      <Drawer placement='bottom' onClose={toggleColorFill} isOpen={isColorFillOpen}>
        <DrawerOverlay />
        <DrawerContent>
          <DrawerHeader borderBottomWidth='1px'>
            <Flex align='center' justify='space-between'>
              <p>Adjust Percentage</p>
              <CloseIcon onClick={toggleColorFill} />
            </Flex>
          </DrawerHeader>
          <DrawerBody>
            <Suspense fallback={<Box />}>
              <FillPercentOption value={fillPercent || 0} onChange={onUpdateFillPercent} />
            </Suspense>
          </DrawerBody>
        </DrawerContent>
      </Drawer>

      <Drawer placement='bottom' onClose={toggleFillColor} isOpen={isFillColorOpen}>
        <DrawerOverlay />
        <DrawerContent>
          <DrawerHeader borderBottomWidth='1px'>
            <Flex align='center' justify='space-between'>
              <p>Fill Color</p>
              <CloseIcon onClick={toggleFillColor} />
            </Flex>
          </DrawerHeader>
          <DrawerBody>
            <Suspense fallback={<Box />}>
              <ColorPickerInContext
                iconStyle='text'
                color={fillColor[0] ?? ''}
                label={'Fill Color'}
                onChange={onUpdatePrimaryFillColor}
                showNoColorOption={false}
              />
            </Suspense>
          </DrawerBody>
        </DrawerContent>
      </Drawer>

      <Drawer placement='bottom' onClose={toggleFillColorTwo} isOpen={isFillColorTwoOpen}>
        <DrawerOverlay />
        <DrawerContent>
          <DrawerHeader borderBottomWidth='1px'>
            <Flex align='center' justify='space-between'>
              <p>Fill Color Two</p>
              <CloseIcon onClick={toggleFillColorTwo} />
            </Flex>
          </DrawerHeader>
          <DrawerBody>
            <Suspense fallback={<Box />}>
              <ColorPickerInContext
                iconStyle='text'
                color={fillColor[1] ?? ''}
                label={'Fill Color Two'}
                onChange={onUpdateSecondaryFillColor}
                showNoColorOption={false}
              />
            </Suspense>
          </DrawerBody>
        </DrawerContent>
      </Drawer>
    </>
  );
};

const MirrorOptions = () => {
  const { updateWidget, mirror = {} } = useWidget<BasicShapeWidgetData>();

  const onUpdateMirrorVertical = (): void => updateWidget({ mirror: { isVertical: !mirror.isVertical } });
  const onUpdateMirrorHorizontal = (): void => updateWidget({ mirror: { isHorizontal: !mirror.isHorizontal } });

  return (
    <>
      <Box mr='5px' p='10px 10px' as='div' style={{ display: 'flex', flexDirection: 'column', alignItems: 'center' }}>
        <IconButton
          mb='10px'
          onClick={onUpdateMirrorHorizontal}
          variant={'ghost'}
          isActive={mirror.isHorizontal}
          size='sm'
          aria-label={'Flip Horizontal'}
          icon={<HorizontalMirrorIcon />}
        />
        <Box w='75px' as='div'>
          <Text fontSize={10} style={{ fontWeight: 700 }} align='center'>
            Flip Horizontal
          </Text>
        </Box>
      </Box>

      <Box mr='5px' p='10px 10px' as='div' style={{ display: 'flex', flexDirection: 'column', alignItems: 'center' }}>
        <IconButton
          mb='10px'
          onClick={onUpdateMirrorVertical}
          variant={'ghost'}
          isActive={mirror.isVertical}
          size='sm'
          aria-label={'Flip Vertical'}
          icon={<VerticalMirrorIcon />}
        />
        <Box w='75px' as='div'>
          <Text fontSize={10} style={{ fontWeight: 700 }} align='center'>
            Flip Vertical
          </Text>
        </Box>
      </Box>
    </>
  );
};

const BorderColorOption = () => {
  const {
    border: { color, width },
    updateWidget,
  } = useWidget<BasicShapeWidgetData>();
  const { isOpen: isBorderColorOpen, onToggle: toggleBorderColor } = useDisclosure();

  const borderColorTriggerProps: DefaultTriggerProps = useMemo(
    () => ({
      onClick: toggleBorderColor,
    }),
    [toggleBorderColor],
  );

  const onUpdateBorderColor = (borderColor: string): void => {
    const newBorder = {
      color: borderColor,
      ...(width === 0 && { width: DEFAULT_BORDER_WIDTH }),
    };
    updateWidget({ border: newBorder });
  };

  return (
    <>
      <Box
        onClick={() => toggleBorderColor()}
        mr='5px'
        p='10px 10px'
        as='div'
        style={{ display: 'flex', flexDirection: 'column', alignItems: 'center' }}
      >
        <DefaultTrigger iconStyle='border' color={color} {...borderColorTriggerProps} />

        <Box w='75px' as='div'>
          <Text fontSize={10} style={{ fontWeight: 700 }} align='center'>
            Border Color
          </Text>
        </Box>
      </Box>

      <Drawer placement='bottom' onClose={toggleBorderColor} isOpen={isBorderColorOpen}>
        <DrawerOverlay />
        <DrawerContent>
          <DrawerHeader borderBottomWidth='1px'>
            <Flex align='center' justify='space-between'>
              <p>Border Color</p>
              <CloseIcon onClick={toggleBorderColor} />
            </Flex>
          </DrawerHeader>
          <DrawerBody>
            <Suspense fallback={<Box />}>
              <ColorPickerInContext
                iconStyle='text'
                color={color ?? ''}
                label={'Border Color'}
                onChange={onUpdateBorderColor}
                showNoColorOption={false}
              />
            </Suspense>
          </DrawerBody>
        </DrawerContent>
      </Drawer>
    </>
  );
  // return <ColorPicker color={color} onChange={onUpdateBorderColor} label={'Border Color'} iconStyle={'border'} />;
};

const BorderStyleOption = () => {
  const {
    border: { style },
    updateWidget,
  } = useWidget<BasicShapeWidgetData>();
  const { isOpen: isBorderStyleOpen, onToggle: toggleBorderStyle } = useDisclosure();

  const onUpdateBorderStyle = (style: any) => updateWidget({ border: { style: style || BorderStyle.Solid } });

  const selectedIndex = BORDER_STYLE_OPTIONS.findIndex((option) => option.value === style);

  const handleSelect = (e: React.SyntheticEvent<HTMLButtonElement>): void => {
    onUpdateBorderStyle(e.currentTarget.dataset.value);
  };

  return (
    <>
      <Box
        onClick={() => toggleBorderStyle()}
        mr='5px'
        p='10px 10px'
        as='div'
        style={{ display: 'flex', flexDirection: 'column', alignItems: 'center' }}
      >
        <IconButton mb='10px' size='sm' aria-label='Border Style' icon={<BorderStyleIcon />} />
        <Box w='75px' as='div'>
          <Text fontSize={10} style={{ fontWeight: 700 }} align='center'>
            Border Style
          </Text>
        </Box>
      </Box>

      <Drawer placement='bottom' onClose={toggleBorderStyle} isOpen={isBorderStyleOpen}>
        <DrawerOverlay />
        <DrawerContent>
          <DrawerHeader borderBottomWidth='1px'>
            <Flex align='center' justify='space-between'>
              <p>Select Border Style</p>
              <CloseIcon onClick={toggleBorderStyle} />
            </Flex>
          </DrawerHeader>
          <DrawerBody>
            <Suspense fallback={<Box />}>
              <Box p={2}>
                {BORDER_STYLE_OPTIONS.map((opt, i) => (
                  <Button
                    leftIcon={opt.icon}
                    onClick={handleSelect}
                    data-value={opt.value}
                    size={'xs'}
                    key={opt.value}
                    isActive={i === selectedIndex}
                    variant={'toolbar-dropdown-item'}
                  >
                    {opt.label}
                  </Button>
                ))}
              </Box>
            </Suspense>
          </DrawerBody>
        </DrawerContent>
      </Drawer>
    </>
  );
};

const BorderWidthOption = () => {
  const {
    border: { width },
    updateWidget,
  } = useWidget<BasicShapeWidgetData>();
  const { isOpen: isBorderWidthOpen, onToggle: toggleBorderWidth } = useDisclosure();

  const onUpdateBorderWidth = (borderWidth: number) => updateWidget({ border: { width: borderWidth } });

  return (
    <>
      <Box
        onClick={() => toggleBorderWidth()}
        mr='5px'
        p='10px 10px'
        as='div'
        style={{ display: 'flex', flexDirection: 'column', alignItems: 'center' }}
      >
        <IconButton mb='10px' size='sm' aria-label='Border Width' icon={<LineWidthIcon />} />
        <Box w='75px' as='div'>
          <Text fontSize={10} style={{ fontWeight: 700 }} align='center'>
            Border Width
          </Text>
        </Box>
      </Box>

      <Drawer placement='bottom' onClose={toggleBorderWidth} isOpen={isBorderWidthOpen}>
        <DrawerOverlay />
        <DrawerContent>
          <DrawerHeader borderBottomWidth='1px'>
            <Flex align='center' justify='space-between'>
              <p>Adjust Border Width</p>
              <CloseIcon onClick={toggleBorderWidth} />
            </Flex>
          </DrawerHeader>
          <DrawerBody>
            <Suspense fallback={<Box />}>
              <SliderNumberInput
                title={'Border Width'}
                value={width}
                onChange={onUpdateBorderWidth}
                suffix={'%'}
                sliderWidth={20}
                containerProps={{ p: 2, pt: 3 }}
              />
            </Suspense>
          </DrawerBody>
        </DrawerContent>
      </Drawer>
    </>
  );

  /*
    <SliderPopover
      value={width}
      icon={<LineWidthIcon />}
      label={'Border Width'}
      onChange={onUpdateBorderWidth}
      suffix={'px'}
      max={20}
      title={'Width'}
    />
  */
};

const FillPercentOption = ({ value, onChange }: { value: number; onChange: (f: number) => void }) => {
  return (
    <SliderNumberInput
      title={'Color Fill'}
      value={value}
      onChange={onChange}
      suffix={'%'}
      sliderWidth={120}
      containerProps={{ p: 2, pt: 3 }}
    />
  );
};

const DefaultTrigger = ({ iconStyle, color, ...defaultTriggerProps }: Props) => {
  const isBorder = iconStyle === 'border';
  return (
    <Flex
      w={TRIGGER_BUTTON_SIZE}
      h={TRIGGER_BUTTON_SIZE}
      mb='10px'
      borderRadius='5px'
      as='button'
      transition='opacity 0.2s'
      _focus={{
        boxShadow: 'var(--vg-shadows-outline)',
        outline: 'none',
      }}
      _hover={{ bg: 'hover.gray' }}
      justifyContent='center'
      alignItems='center'
      background='#dbdbdb'
      {...defaultTriggerProps}
    >
      <Flex
        w={COLOR_ICON_SIZE}
        h={COLOR_ICON_SIZE}
        borderRadius='sm'
        background={color}
        border='1px solid var(--vg-colors-black-alpha-500)'
        justifyContent='center'
        alignItems='center'
      >
        {isBorder && <Box w={2} h={2} bg='white' boxShadow='0px 0px 0px 1px var(--vg-colors-black-alpha-500) inset' />}
      </Flex>
    </Flex>
  );
};
