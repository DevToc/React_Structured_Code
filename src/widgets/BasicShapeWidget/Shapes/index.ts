import Rectangle, { resizeRectangle } from './Rectangle';
import Ellipse, { resizeEllipse } from './Ellipse';
import SemiCircle, { resizeSemiCircle } from './SemiCircle';
import Triangle, { resizeTriangle } from './Triangle';
import RightTriangle, { resizeRightTriangle } from './RightTriangle';

import { BasicShapeType } from '../BasicShapeWidget.types';

export const SHAPE_TYPE_MAP = {
  [BasicShapeType.Rectangle]: Rectangle,
  [BasicShapeType.Ellipse]: Ellipse,
  [BasicShapeType.SemiCircle]: SemiCircle,
  [BasicShapeType.Triangle]: Triangle,
  [BasicShapeType.RightTriangle]: RightTriangle,
};

export const SHAPE_RESIZE_MAP = {
  [BasicShapeType.Rectangle]: resizeRectangle,
  [BasicShapeType.Ellipse]: resizeEllipse,
  [BasicShapeType.SemiCircle]: resizeSemiCircle,
  [BasicShapeType.Triangle]: resizeTriangle,
  [BasicShapeType.RightTriangle]: resizeRightTriangle,
};
