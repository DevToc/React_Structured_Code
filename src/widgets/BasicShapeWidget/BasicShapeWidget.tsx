import { ReactElement, useState, useCallback, useEffect, Suspense } from 'react';
import { Box, Spinner } from '@chakra-ui/react';

import { CustomOnResize } from 'modules/Editor/components/BoundingBox/BoundingBox.types';
import { useWidget, WidgetBase, WidgetBaseProp, ReadOnlyWidgetBase, WidgetToolbar, TransparentImg } from 'widgets/sdk';
import { BasicShapeLayoutProps, BasicShapeToolbarMenuProps, BasicShapeWidgetData } from './BasicShapeWidget.types';
import { SHAPE_TYPE_MAP, SHAPE_RESIZE_MAP } from './Shapes';
import { generateShapeStyle } from './BasicShapeWidget.helpers';
import { useDynamicImport } from 'hooks/useDynamicImport';
import { WidgetComponentsMap } from 'widgets/components.lazy';

const BasicShapeWidgetToolbarMenu = (props: BasicShapeToolbarMenuProps): ReactElement => {
  const Component = useDynamicImport('basicShapeWidgetToolbar', WidgetComponentsMap);

  return (
    <Suspense fallback={<Spinner />}>
      <Component {...props} />
    </Suspense>
  );
};

export const BasicShapeWidget = (): ReactElement => {
  const { widthPx, heightPx, border, type, widgetId } = useWidget<BasicShapeWidgetData>();
  const [dimension, setDimension] = useState({ width: widthPx, height: heightPx });

  useEffect(() => {
    setDimension({ width: widthPx, height: heightPx });
  }, [widthPx, heightPx]);

  /**
   * Custom onResize handler to update the shape dimensions
   * to allow for live rendering
   */
  const customOnResize = useCallback(
    ({ event, onResize }: CustomOnResize) => {
      onResize(event);

      if (SHAPE_RESIZE_MAP[type]) {
        const widthPx = event.target.clientWidth;
        const heightPx = event.target.clientHeight;

        // resize the shape without setting the new dimensions to the state
        SHAPE_RESIZE_MAP[type]({ widthPx, heightPx, borderWidth: border.width, widgetId });
      }
    },
    [border.width, type, widgetId],
  );

  return (
    <WidgetBase onResize={customOnResize}>
      <WidgetToolbar>
        <BasicShapeWidgetToolbarMenu />
      </WidgetToolbar>
      <BasicShapeLayout dimension={dimension} />
    </WidgetBase>
  );
};

export const ReadOnlyBasicShapeWidget = ({ includeAltTextImg }: WidgetBaseProp): ReactElement => {
  const { widthPx, heightPx } = useWidget<BasicShapeWidgetData>();
  const dimension = { width: widthPx, height: heightPx };

  return (
    <ReadOnlyWidgetBase>
      <BasicShapeLayout dimension={dimension} />
      {includeAltTextImg && <TransparentImg />}
    </ReadOnlyWidgetBase>
  );
};

const BasicShapeLayout = ({ dimension }: BasicShapeLayoutProps): ReactElement => {
  const { type, border, fillColor, fillPercent, mirror, cornerRadius, widgetId } = useWidget<BasicShapeWidgetData>();
  const { width, height } = dimension;
  const Shape = SHAPE_TYPE_MAP[type];

  const style = generateShapeStyle(mirror);

  return (
    <Box w={'100%'} h={'100%'} position={'absolute'}>
      <Shape
        widgetId={widgetId}
        fillColor={fillColor}
        fillPercent={fillPercent}
        widthPx={width}
        heightPx={height}
        border={border}
        style={style}
        {...(cornerRadius && { cornerRadius })}
      />
    </Box>
  );
};
