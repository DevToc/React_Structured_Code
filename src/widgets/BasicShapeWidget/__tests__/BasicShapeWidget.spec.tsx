import { renderWidget } from 'utils/test-utils.test';
import { BasicShapeWidget } from '../BasicShapeWidget';
import { generateDefaultData } from '../BasicShapeWidget.helpers';
import { BasicShapeType } from '../BasicShapeWidget.types';

describe('widgets/BasicShapeWidget', () => {
  it('should render correctly with fill', () => {
    const widgetId = '006.widget-1';
    const widget = { widgetId, ...generateDefaultData(BasicShapeType.Rectangle, false) };

    const { asFragment } = renderWidget(<BasicShapeWidget />, { widget });
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render correctly with border', () => {
    const widgetId = '006.widget-1';
    const widget = { widgetId, ...generateDefaultData(BasicShapeType.Rectangle, true) };
    const { asFragment } = renderWidget(<BasicShapeWidget />, { widget });

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render rounded rectangle', () => {
    const widgetId = '006-widget-1';
    const widget = { widgetId, ...generateDefaultData(BasicShapeType.Rectangle, false, { cornerRadius: 10 }) };
    const { asFragment } = renderWidget(<BasicShapeWidget />, { widget });

    expect(asFragment()).toMatchSnapshot();
  });
});
