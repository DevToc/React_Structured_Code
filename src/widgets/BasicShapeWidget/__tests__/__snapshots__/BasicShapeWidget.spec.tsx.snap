// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`widgets/BasicShapeWidget should render correctly with border 1`] = `
<DocumentFragment>
  <div
    aria-describedby="006.widget-1-description"
    aria-label="Basic shape widget"
    class="widget-container widget-selecto-target  css-1eee8vp"
    data-testid="widgetbase-006.widget-1"
    id="006.widget-1"
    style="display: block; position: absolute; top: 0px; left: 0px; width: 100px; height: 100px; transform: rotate(0deg); z-index: 0; cursor: move;"
    tabindex="-1"
  >
    <div
      class="css-dh3it3"
    >
      <svg
        height="100px"
        id="triangle-svg-006.widget-1"
        style="position: absolute;"
        width="100px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="transparent"
          height="90px"
          id="triangle-fill-006.widget-1"
          rx="0px"
          width="90px"
          x="5"
          y="5"
        />
        <rect
          fill="none"
          height="95px"
          id="triangle-border-006.widget-1"
          rx="0px"
          stroke="#0073E6"
          stroke-width="5px"
          width="95px"
          x="2.5"
          y="2.5"
        />
      </svg>
    </div>
    <span
      class="css-f8n5zr"
    >
      <p
        class="chakra-text css-0"
        id="006.widget-1-description"
      />
    </span>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`widgets/BasicShapeWidget should render correctly with fill 1`] = `
<DocumentFragment>
  <div
    aria-describedby="006.widget-1-description"
    aria-label="Basic shape widget"
    class="widget-container widget-selecto-target  css-1eee8vp"
    data-testid="widgetbase-006.widget-1"
    id="006.widget-1"
    style="display: block; position: absolute; top: 0px; left: 0px; width: 100px; height: 100px; transform: rotate(0deg); z-index: 0; cursor: move;"
    tabindex="-1"
  >
    <div
      class="css-dh3it3"
    >
      <svg
        height="100px"
        id="triangle-svg-006.widget-1"
        style="position: absolute;"
        width="100px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="#0073E6"
          height="100px"
          id="triangle-fill-006.widget-1"
          rx="0px"
          width="100px"
          x="0"
          y="0"
        />
        <rect
          fill="none"
          height="100px"
          id="triangle-border-006.widget-1"
          rx="0px"
          stroke="black"
          stroke-width="0px"
          width="100px"
          x="0"
          y="0"
        />
      </svg>
    </div>
    <span
      class="css-f8n5zr"
    >
      <p
        class="chakra-text css-0"
        id="006.widget-1-description"
      />
    </span>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`widgets/BasicShapeWidget should render rounded rectangle 1`] = `
<DocumentFragment>
  <div
    aria-describedby="006-widget-1-description"
    aria-label="Basic shape widget"
    class="widget-container widget-selecto-target  css-1eee8vp"
    data-testid="widgetbase-006-widget-1"
    id="006-widget-1"
    style="display: block; position: absolute; top: 0px; left: 0px; width: 100px; height: 100px; transform: rotate(0deg); z-index: 0; cursor: move;"
    tabindex="-1"
  >
    <div
      class="css-dh3it3"
    >
      <svg
        height="100px"
        id="triangle-svg-006-widget-1"
        style="position: absolute;"
        width="100px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="#0073E6"
          height="100px"
          id="triangle-fill-006-widget-1"
          rx="10px"
          width="100px"
          x="0"
          y="0"
        />
        <rect
          fill="#0073E6"
          height="100px"
          id="triangle-border-006-widget-1"
          rx="10px"
          stroke="black"
          stroke-width="0px"
          width="100px"
          x="0"
          y="0"
        />
      </svg>
    </div>
    <span
      class="css-f8n5zr"
    >
      <p
        class="chakra-text css-0"
        id="006-widget-1-description"
      />
    </span>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
